drivers/mpu9250/MPU9250_mag.cpp

```diff
@@ -0,0 +1,462 @@
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "MPU9250.hpp"
+#include "MPU9250_mag.hpp"
+
+using namespace DriverFramework;
+
+// Uncomment to allow additional debug output to be generated.
+// #define MPU9250_MAG_DEBUG 1
+
+int MPU9250_mag::_convert_sample_rate_enum_to_hz(
+		enum mag_sample_rate_e sample_rate)
+{
+	switch (sample_rate) {
+		case MPU9250_MAG_SAMPLE_RATE_100HZ:
+			return 100;
+		case MPU9250_MAG_SAMPLE_RATE_8HZ:
+			return 8;
+		default:
+			DF_LOG_ERR("Invalid mag sample rate detected.");
+			return -1;
+	}
+}
+
+int MPU9250_mag::_initialize(int output_data_rate_in_hz)
+{
+	int result;
+
+	// Configure the IMU as an I2C master at 400 KHz.
+	result = _imu.writeReg(MPUREG_USER_CTRL, BITS_USER_CTRL_I2C_MST_EN);
+	if (result != 0) {
+		DF_LOG_ERR("IMU I2C master enable failed.");
+		return -1;
+	}
+	result = _imu.writeReg(MPUREG_I2C_MST_CTRL, BITS_I2C_MST_CLK_400_KHZ);
+	if (result != 0) {
+		DF_LOG_ERR("IMU I2C master bus config failed.");
+		return -1;
+	}
+	usleep(1000);
+
+	// Detect mag presence by reading whoami register
+	if (detect() != 0) {
+		DF_LOG_ERR("MPU9250 mag not detected.");
+		return -1;
+	}
+
+	// Get mag calibraion data from Fuse ROM
+	if (get_sensitivity_adjustment() != 0) {
+		DF_LOG_ERR("Unable to read mag sensitivity adjustment");
+		return -1;
+	}
+
+	// Power on and configure the mag to produce 16 bit data in continuous measurement mode.
+	int mag_mode;
+	if (_sample_rate == MPU9250_MAG_SAMPLE_RATE_100HZ) {
+		mag_mode = BIT_MAG_CNTL1_MODE_CONTINUOUS_MEASURE_MODE_2;
+	} else if (_sample_rate == MPU9250_MAG_SAMPLE_RATE_8HZ) {
+		mag_mode = BIT_MAG_CNTL1_MODE_CONTINUOUS_MEASURE_MODE_1;
+	} else {
+		DF_LOG_ERR("Unable to select a valid mag mode.");
+		return -1;
+	}
+	result = write_reg(MPU9250_MAG_REG_CNTL1, BIT_MAG_CNTL1_16_BITS | mag_mode);
+	if (result != 0) {
+		DF_LOG_ERR("Unable to configure the magnetometer mode.");
+	}
+	usleep(1000);
+
+	// Slave 0 provides ST1, mag data, and ST2 data in a bulk transfer of
+	// 8 bytes of data.  Use the address of ST1 in SLV0_REG as the beginning
+	// register of the 8 byte bulk transfer.
+	result = _imu.writeReg(MPUREG_I2C_SLV0_ADDR,
+			MPU9250_AK8963_I2C_ADDR | MPU9250_AK8963_I2C_READ);
+	if (result != 0) {
+		DF_LOG_ERR("MPU9250 I2C slave 0 address configuration failed.");
+		return -1;
+	}
+	result = _imu.writeReg(MPUREG_I2C_SLV0_REG, MPU9250_MAG_REG_ST1);
+	if (result != 0) {
+		DF_LOG_ERR("MPU9250 I2C slave 0 register configuration failed.");
+		return -1;
+	}
+	result = _imu.writeReg(MPUREG_I2C_SLV0_CTRL, BITS_I2C_SLV0_EN | 0x08);
```

Comment:
@jywilson MPU9250_AK8963_I2C_READ == 0x8, which is == 0x08. Your statement is not correct as 0x08 != 0x80. Please do not use magic numbers and proper defines instead.